teku/src/main/java/tech/pegasys/teku/cli/subcommand/RemoteSpecLoader.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2021 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.cli.subcommand;
+
+import java.net.URI;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import tech.pegasys.teku.infrastructure.exceptions.InvalidConfigurationException;
+import tech.pegasys.teku.infrastructure.logging.SubCommandLogger;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.SpecFactory;
+import tech.pegasys.teku.spec.config.SpecConfigLoader;
+import tech.pegasys.teku.validator.remote.apiclient.OkHttpClientAuthLoggingIntercepter;
+import tech.pegasys.teku.validator.remote.apiclient.OkHttpValidatorRestApiClient;
+
+class RemoteSpecLoader {
+
+  private static final long RETRY_DELAY = 5000;
+
+  static Spec getSpec(OkHttpValidatorRestApiClient apiClient) {
+    try {
+      return apiClient
+          .getConfigSpec()
+          .map(response -> SpecConfigLoader.loadConfig(response.data))
+          .map(SpecFactory::create)
+          .orElseThrow();
+    } catch (Exception e) {
+      String errMsg =
+          String.format(
+              "Failed to retrieve network spec from beacon node endpoint '%s'.\nDetails: %s",
+              apiClient.getBaseEndpoint(), e.getMessage());
+      throw new InvalidConfigurationException(errMsg, e);
+    }
+  }
+
+  static Spec getSpec(URI beaconEndpoint) {
+    return getSpec(createApiClient(beaconEndpoint));
+  }
+
+  static Spec getSpecWithRetry(URI beaconEndpoint) {
+    return retry(() -> getSpec(beaconEndpoint));
+  }
+
+  static Spec getSpecWithRetry(OkHttpValidatorRestApiClient apiClient) {
+    return retry(() -> getSpec(apiClient));
+  }
+
+  private static <T> T retry(Callable<T> f) {
+    try {
+      return f.call();
+    } catch (Throwable e) {
+      logError(e);
+      sleep();
+      return retry(f);
+    }
+  }
+
+  private static void sleep() {
+    try {
+      Thread.sleep(RETRY_DELAY);
+    } catch (InterruptedException e) {
+      logError(e);
+      throw new RuntimeException(e);
+    }
```

Comment:
InterruptedException won't have any message, so it seems to mostly just print 'null' on the console.

The main case we'd be likely to be interrupted in this sleep would be if we're starting up, and getting the 'failed to retrieve network spec' message, and ctrl+c, in which case we know what's causing the issue I guess.
Could potentially just `throw new ShuttingDownException()` here.

Reply:
makes sense. done